sql: update sql activity update job transaction queries

xinhaoz opened the pull request:
This commit updates the queries used to populate the
`system.transaction_activity` table to simplify some
of the logic and to use AOST follower reads. Due to
the addition of AOST, the queries have been split up
to a distinct SELECT and UPSERT query, rather than a
single upsert using a select.

Part of: https://github.com/cockroachdb/cockroach/issues/133495
Epic: none
Release note: None


Commit 2:
### sql: refactor hardcoded query plan testing for txn activity
Previously we had logic tests in `observability` that tested
the query plans produced by the queries in the sql activity
update job.

This commit moves the queries related from the transactions
table over to a new file `sql_activity_job_queries` where
it is tested via custom data driven test commands that directly
call the function that constructs each query.

Part of: https://github.com/cockroachdb/cockroach/issues/133495
Epic: none
Release note: None

xinhaoz committed:
sql: update sql activity update job transaction queries

This commit updates the queries used to populate the
`system.transaction_activity` table to simplify some
of the logic and to use AOST follower reads. Due to
the addition of AOST, the queries have been split up
to a distinct SELECT and UPSERT query, rather than a
single upsert using a select.

Part of: #133495
Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143364)
<!-- Reviewable:end -->

xinhaoz committed:
sql: refactor hardcoded query plan testing for txn activity

Previously we had logic tests in `observability` that tested
the query plans produced by the queries in the sql activity
update job.

This commit moves the queries related from the transactions
table over to a new file `sql_activity_job_queries` where
it is tested via custom data driven test commands that directly
call the function that constructs each query.

Part of: #133495
Epic: none
Release note: None

xinhaoz commented:
To explain the long line count: the second commit is mostly a testing overhaul for some hard coded query plan testing we were doing in logic tests.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice! I have limited experience with these queries, so definitely get someone else to sign-off.

> Reviewed 2 of 2 files at r1, 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143364#-:-OMDvkqo-3CW5njR7yH3:bcl8zjx)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons and @michae2)

___
*[`-- commits` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/143364#-OMDvdSS0MV5OBgBiNqF:-OMDvdSS0MV5OBgBiNqG:b-j124fb):*
Great idea! This will help us catch query plan regressions.
___
*[`pkg/sql/sql_activity_update_job.go` line 400 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143364#-OMDvSlyB9P8d2SrsX8f:-OMDvSlz5Ph2Ng9M-7Ba:bx8xoqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d0ca19a47db3d8b0b69bb89e4c98e2473e4053d/pkg/sql/sql_activity_update_job.go#L400)):*
> ```Go
> 	//
> 	// Note that it's important we use follower reads here to avoid causing
> 	// contention // since the rest of the cluster is still writing to the statistics
> ```

nit: remove extra `//`


<!-- Sent from Reviewable.io -->

angles-n-daemons commented on pkg/sql/sql_activity_update_job.go:
```diff
@@ -762,3 +701,207 @@ func (u *sqlActivityUpdater) getTableRowCount(
+       COALESCE((statistics->'execution_statistics'->'cpuSQLNanos'->>'mean')::float,0),
+       (statistics->'statistics'->'svcLat'->>'mean')::float
+     FROM (
+       SELECT max(aggregated_ts) AS max_aggregated_ts,
```
Might be worth changing this to a CTE as well if it's not too much trouble, eg:

```sql
WITH aggregate_transaction_statistics AS (
    SELECT max(aggregated_ts) AS max_aggregated_ts,
    ...
)

xinhaoz replied:
Done!

angles-n-daemons commented on pkg/sql/sql_activity_update_job.go:
```diff
@@ -762,3 +701,207 @@ func (u *sqlActivityUpdater) getTableRowCount(
+       WHERE aggregated_ts = $1 AND
+             app_name not like '$ internal%%'
+       GROUP BY app_name, fingerprint_id
+     ) %[1]s
```
does an aost clause do anything on a subquery?

xinhaoz replied:
It's just syntatically required for aost to work.

angles-n-daemons commented on pkg/sql/sql_activity_update_job.go:
```diff
@@ -382,9 +363,39 @@ func (u *sqlActivityUpdater) transferTopStats(
 	totalEstimatedTxnClusterExecSeconds float64,
 ) (retErr error) {
 
+	// Select the top transactions for each of the following 5 columns in the
```
Nice, this is a great comment - thanks for that.

angles-n-daemons reviewed (commented):
Looks great! Couple of questions on the changeset, overall looks good.

xinhaoz commented on pkg/sql/sql_activity_update_job.go:
```diff
@@ -762,3 +701,207 @@ func (u *sqlActivityUpdater) getTableRowCount(
+        (merge_stats -> 'statistics' -> 'svcLat' ->> 'mean')::float AS svc_lat,
+        COALESCE((merge_stats -> 'execution_statistics' -> 'contentionTime' ->> 'mean')::float, 0) AS contention_time,
+        COALESCE((merge_stats -> 'execution_statistics' -> 'cpuSQLNanos' ->> 'mean')::float, 0) AS cpu_time
+    FROM transaction_aggregates %[1]s
```
Same here - just required in order to use AOST for the entire query.

xinhaoz commented on pkg/sql/sql_activity_update_job.go:
```diff
@@ -762,3 +701,207 @@ func (u *sqlActivityUpdater) getTableRowCount(
+			}
+		}
+
+		_, err := executor.ExecEx(ctx, redact.RedactableString(opName),
```
This is just the type required by the function -- from my understanding, since we have no redaction markers it'll print as is. I've swapped to use `redact.Sprintf` here.

angles-n-daemons replied:
Ah thanks for the additional information 👍

xinhaoz commented:
tftr!
bors r+
